File: mpi/S2-ref/1d2.c
/*
//   1D Steady-State Heat Transfer 
//   FEM with Piece-wise Linear Elements
//   CG (Conjugate Gradient) Method 
//
//   d/dx(CdT/dx) + Q = 0
//   T=0@x=0  
*/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>
#include <mpi.h>

int main(int argc, char **argv){

  int NE, N, NP, NPLU, IterMax, errno;
  int NEg, Ng;

  double dX, Resid, Eps, Area, QV, COND, QN;
  double X1, X2, DL, Ck;
  double *PHI, *Rhs, *X;
  double *Diag, *AMat;
  double *R, *Z, *Q, *P, *DD;

  int *Index, *Item, *Icelnod;
  double Kmat[2][2], Emat[2][2];

  int i, j, in1, in2, k, icel, k1, k2, jS;
  int iter, nr, neib;
  FILE *fp;
  double BNorm2, Rho, Rho1=0.0, C1, Alpha, Beta, DNorm2;

  int PETot, MyRank, kk, is, ir, len_s, len_r, tag;
  int NeibPETot, BufLength;
  int NeibPE[2];

  int import_index[3], import_item[2];
  int export_index[3], export_item[2];
  double SendBuf[2], RecvBuf[2];

  double BNorm20, Rho0, C10, DNorm20;
  double S1Time, E1Time;
  double S2Time, E2Time;

  MPI_Status  *StatSend,    *StatRecv;
  MPI_Request *RequestSend, *RequestRecv;

  int ierr = 1;

/*
// +-------+ 
// | INIT. |
// +-------+ 
//=== */

/*
//-- CONTROL data
*/

    ierr = MPI_Init(&argc, &argv);
    ierr = MPI_Comm_size(MPI_COMM_WORLD, &PETot);
    ierr = MPI_Comm_rank(MPI_COMM_WORLD, &MyRank);

    if(MyRank == 0){
      fp = fopen("input.dat", "r");
      assert(fp != NULL);
      fscanf(fp, "%d", &NEg);
      fscanf(fp, "%lf %lf %lf %lf", &dX, &QV, &Area, &COND);
      fscanf(fp, "%d", &IterMax);
      fscanf(fp, "%lf", &Eps);
      fclose(fp);
    }

    ierr = MPI_Bcast(&NEg    , 1, MPI_INT, 0, MPI_COMM_WORLD);
    ierr = MPI_Bcast(&IterMax, 1, MPI_INT, 0, MPI_COMM_WORLD);
    ierr = MPI_Bcast(&dX     , 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    ierr = MPI_Bcast(&QV     , 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    ierr = MPI_Bcast(&Area   , 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    ierr = MPI_Bcast(&COND   , 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
    ierr = MPI_Bcast(&Eps    , 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);

/*
//-- LOCAL MESH size
*/
    Ng= NEg + 1;
    N = Ng / PETot;

    nr = Ng - N*PETot;
    if(MyRank < nr) N++;

    NE= N - 1 + 2;
    NP= N + 2;
    if(MyRank == 0) NE= N - 1 + 1;
    if(MyRank == 0) NP= N + 1;
    if(MyRank == PETot-1) NE= N - 1 + 1;
    if(MyRank == PETot-1) NP= N + 1;

    if(PETot==1){NE=N-1;}
    if(PETot==1){NP=N  ;}

/*
/-- Arrays
*/
    PHI  = calloc(NP, sizeof(double));
    Diag = calloc(NP, sizeof(double));
    AMat = calloc(2*NP-2, sizeof(double));
    Rhs  = calloc(NP, sizeof(double));
    Index= calloc(NP+1, sizeof(int));
    Item = calloc(2*NP-2, sizeof(int));
    Icelnod= calloc(2*NE, sizeof(int));

    for(i=0;i<NP;i++)  PHI[i] = 0.0;
    for(i=0;i<NP;i++) Diag[i] = 0.0;
    for(i=0;i<NP;i++)  Rhs[i] = 0.0;
    for(k=0;k<2*NP-2;k++)  AMat[k] = 0.0;

    for(i=0;i<3;i++) import_index[i]= 0;
    for(i=0;i<3;i++) export_index[i]= 0;
    for(i=0;i<2;i++) import_item[i]= 0;
    for(i=0;i<2;i++) export_item[i]= 0;

    for(icel=0;icel<NE;icel++){
      Icelnod[2*icel  ]= icel;     
      Icelnod[2*icel+1]= icel+1;     
    }

    if(PETot>1){
      if(MyRank==0){
	icel= NE-1;
	Icelnod[2*icel  ]= N-1;     
	Icelnod[2*icel+1]= N;
      }else if(MyRank==PETot-1){
	icel= NE-1;
	Icelnod[2*icel  ]= N;
	Icelnod[2*icel+1]= 0;
      }else{ 
	icel= NE-2;
	Icelnod[2*icel  ]= N;
	Icelnod[2*icel+1]= 0;
	icel= NE-1;
	Icelnod[2*icel  ]= N-1;
	Icelnod[2*icel+1]= N+1;
      }
    }

    Kmat[0][0]= +1.0;
    Kmat[0][1]= -1.0;
    Kmat[1][0]= -1.0;
    Kmat[1][1]= +1.0;

/*
// +--------------+
// | CONNECTIVITY |
// +--------------+
*/
    for(i=0;i<N+1;i++)    Index[i] = 2;
    for(i=N+1;i<NP+1;i++) Index[i] = 1;

    Index[0] = 0;
    if(MyRank == 0)       Index[1] = 1;
    if(MyRank == PETot-1) Index[N] = 1;

    for(i=0;i<NP;i++){
      Index[i+1]= Index[i+1] + Index[i];     
    }
    
    NPLU= Index[NP];

    for(i=0;i<N;i++){
      jS = Index[i];
      if((MyRank==0)&&(i==0)){
	Item[jS] = i+1;
      }else if((MyRank==PETot-1)&&(i==N-1)){
	Item[jS] = i-1;
      }else{
	Item[jS]   = i-1;
	Item[jS+1] = i+1;
	if(i==0)  { Item[jS]  = N;}
	if(i==N-1){ Item[jS+1]= N+1;}
	if((MyRank==0)&&(i==N-1)){Item[jS+1]= N;}
      }
    }

    i =N;
    jS= Index[i];
    if (MyRank==0) {
      Item[jS]= N-1;
    }else {
      Item[jS]= 0;
    }

    i =N+1;
    jS= Index[i];
    if ((MyRank!=0)&&(MyRank!=PETot-1)) {
      Item[jS]= N-1;
    }

/*
//-- COMMUNICATION
*/
    NeibPETot = 2;
    if(MyRank == 0)       NeibPETot = 1;
    if(MyRank == PETot-1) NeibPETot = 1;
    if(PETot  == 1)       NeibPETot = 0;

    NeibPE[0] = MyRank - 1;
    NeibPE[1] = MyRank + 1;

    if(MyRank == 0)     NeibPE[0] = MyRank + 1;
    if(MyRank == PETot-1) NeibPE[0] = MyRank - 1;

    import_index[1]=1;
    import_index[2]=2;
    import_item[0]=  N;
    import_item[1]=  N+1;

    export_index[1]=1;
    export_index[2]=2;
    export_item[0]= 0;
    export_item[1]= N-1;

    if(MyRank == 0) import_item[0]=N;
    if(MyRank == 0) export_item[0]=N-1;

    BufLength = 1;


    StatSend = malloc(sizeof(MPI_Status) * 2*NeibPETot);
    RequestSend = malloc(sizeof(MPI_Request) * 2*NeibPETot);

    ierr = MPI_Barrier(MPI_COMM_WORLD);
    S1Time = MPI_Wtime();
/*
// +-----------------+
// | MATRIX assemble |
// +-----------------+
*/
    for(icel=0;icel<NE;icel++){
      in1= Icelnod[2*icel];
      in2= Icelnod[2*icel+1];
      DL = dX;

      Ck= Area*COND/DL;
      Emat[0][0]= Ck*Kmat[0][0];
      Emat[0][1]= Ck*Kmat[0][1];
      Emat[1][0]= Ck*Kmat[1][0];
      Emat[1][1]= Ck*Kmat[1][1];

      Diag[in1]= Diag[in1] + Emat[0][0];
      Diag[in2]= Diag[in2] + Emat[1][1];

      if ((MyRank==0)&&(icel==0)){
	k1=Index[in1];
      }else {k1=Index[in1]+1;}

      k2=Index[in2];

      AMat[k1]= AMat[k1] + Emat[0][1];
      AMat[k2]= AMat[k2] + Emat[1][0];

      QN= 0.5*QV*Area*dX;
      Rhs[in1]= Rhs[in1] + QN;
      Rhs[in2]= Rhs[in2] + QN;
    }

/*
// +---------------------+
// | BOUNDARY conditions |
// +---------------------+
*/

/* X=Xmin */
    if (MyRank==0){      
      i=0;
      jS= Index[i];
      AMat[jS]= 0.0;
      Diag[i ]= 1.0;
      Rhs [i ]= 0.0;

      for(k=0;k<NPLU;k++){
	if(Item[k]==0){AMat[k]=0.0;}
      }
    }
    E1Time = MPI_Wtime();


/*
// +---------------+
// | CG iterations |
// +---------------+
//=== */
    R = calloc(NP, sizeof(double));
    Z = calloc(NP, sizeof(double));
    P = calloc(NP, sizeof(double));
    Q = calloc(NP, sizeof(double));
    DD= calloc(NP, sizeof(double));

    for(i=0;i<N;i++){
      DD[i]= 1.0 / Diag[i];
    }

/*
//-- {r0}= {b} - [A]{xini} |
*/
    for(neib=0;neib<NeibPETot;neib++){
      for(k=export_index[neib];k<export_index[neib+1];k++){
	kk= export_item[k];
	SendBuf[k]= PHI[kk];
      }
    }

    for(neib=0;neib<NeibPETot;neib++){
      is   = export_index[neib];
      len_s= export_index[neib+1] - export_index[neib];
      MPI_Isend(&SendBuf[is], len_s, MPI_DOUBLE, NeibPE[neib],
		0, MPI_COMM_WORLD, &RequestSend[neib]);
    }
      
    for(neib=0;neib<NeibPETot;neib++){
      ir   = import_index[neib];
      len_r= import_index[neib+1] - import_index[neib];
      MPI_Irecv(&PHI[ir+N], len_r, MPI_DOUBLE, NeibPE[neib],
		0, MPI_COMM_WORLD, &RequestSend[neib+NeibPETot]);
    }
    MPI_Waitall(2*NeibPETot, RequestSend, StatSend);

    for(i=0;i<N;i++){
      R[i] = Diag[i]*PHI[i];
      for(j=Index[i];j<Index[i+1];j++){
	R[i] += AMat[j]*PHI[Item[j]];
      }
    }

    BNorm20 = 0.0;
    for(i=0;i<N;i++){
      BNorm20 += Rhs[i] * Rhs[i];
      R[i] = Rhs[i] - R[i];
    }
    
    ierr = MPI_Allreduce(&BNorm20, &BNorm2, 1, MPI_DOUBLE, 
			 MPI_SUM, MPI_COMM_WORLD);

/* /******************************************************************** */
    for(iter=1;iter<=IterMax;iter++){
/*
//-- {z}= [Minv]{r}
*/
    for(i=0;i<N;i++){
      Z[i] = DD[i] * R[i];
    }

/*
//-- RHO= {r}{z}
*/
    Rho0= 0.0;
    for(i=0;i<N;i++){
      Rho0 += R[i] * Z[i];
    }
    ierr = MPI_Allreduce(&Rho0, &Rho, 1, MPI_DOUBLE, 
			 MPI_SUM, MPI_COMM_WORLD);

/*
//-- {p} = {z} if      ITER=1  
//   BETA= RHO / RHO1  otherwise 
*/
    if(iter == 1){
      for(i=0;i<N;i++){
	P[i] = Z[i];
      }
    }else{
      Beta = Rho / Rho1;
      for(i=0;i<N;i++){
	P[i] = Z[i] + Beta*P[i];
      }
    }

/*
//-- {q}= [A]{p}
*/
    for(neib=0;neib<NeibPETot;neib++){
      for(k=export_index[neib];k<export_index[neib+1];k++){
	kk= export_item[k];
	SendBuf[k]= P[kk];
      }
    }

    for(neib=0;neib<NeibPETot;neib++){
      is   = export_index[neib];
      len_s= export_index[neib+1] - export_index[neib];
      MPI_Isend(&SendBuf[is], len_s, MPI_DOUBLE, NeibPE[neib],
		0, MPI_COMM_WORLD, &RequestSend[neib]);
    }

    for(neib=0;neib<NeibPETot;neib++){
      ir   = import_index[neib];
      len_r= import_index[neib+1] - import_index[neib];
      MPI_Irecv(&P[ir+N], len_r, MPI_DOUBLE, NeibPE[neib],
		0, MPI_COMM_WORLD, &RequestSend[neib+NeibPETot]);
    }
        
    MPI_Waitall(2*NeibPETot, RequestSend, StatSend);
    
    for(i=0;i<N;i++){
      Q[i] = Diag[i] * P[i];
      for(j=Index[i];j<Index[i+1];j++){
	Q[i] += AMat[j]*P[Item[j]];
      }
    }

/*
//-- ALPHA= RHO / {p}{q}
*/
    C10 = 0.0;
    for(i=0;i<N;i++){
      C10 += P[i] * Q[i];
    }
    
    ierr = MPI_Allreduce(&C10, &C1, 1, MPI_DOUBLE, 
			   MPI_SUM, MPI_COMM_WORLD);
    Alpha = Rho / C1;

/*
//-- {x}= {x} + ALPHA*{p}
//   {r}= {r} - ALPHA*{q}
*/
    for(i=0;i<N;i++){
      PHI[i] += Alpha * P[i];
      R[i] -= Alpha * Q[i];
    }
    
    DNorm20 = 0.0;
    for(i=0;i<N;i++){
      DNorm20 += R[i] * R[i];
    }

    ierr = MPI_Allreduce(&DNorm20, &DNorm2, 1, MPI_DOUBLE, 
			 MPI_SUM, MPI_COMM_WORLD);

    Resid = sqrt(DNorm2/BNorm2);
/*    if (MyRank==0) 
      printf("%8d%s%16.6e\n", iter, " iters, RESID=", Resid);
*/    
    if(Resid <= Eps){
      ierr = 0;
      break;
    }

    Rho1 = Rho;

    }
/* /******************************************************************** */

    E2Time = MPI_Wtime();

/* ********************************************************************
*/

/*
//-- OUTPUT
*/
    if (MyRank==0) 
      printf("%8d%s%16.6e\n", iter, " iters, RESID=", Resid);

    if (MyRank==0) 
      printf("%16.6e%16.6e\n", E1Time-S1Time, E2Time-E1Time);

    if (MyRank==PETot-1) {
      printf("\n%s\n", "### TEMPERATURE");
      printf("%3d%8d%27.20e\n\n\n", MyRank, N, PHI[N-1]);
    }

    ierr = MPI_Finalize();
    return ierr;
}

File: mpi/S2-ref/1d2.f
!C
!C   1D Steady-State Heat Transfer 
!C   FEM with Piece-wise Linear Elements
!C   CG (Conjugate Gradient) Method 
!C
!C   d/dx(CdT/dx) + Q = 0
!C   T=0@x=0  
!C
      program heat1Dp
      implicit REAL*8 (A-H,O-Z)
      include 'mpif.h'

      integer :: N, NPLU, ITERmax
      integer :: R, Z, P, Q, DD

      real(kind=8) :: dX, RESID, EPS 
      real(kind=8) :: AREA, QV, COND
      real(kind=8), dimension(:), allocatable :: PHI, RHS
      real(kind=8), dimension(:  ), allocatable :: DIAG, AMAT
      real(kind=8), dimension(:,:), allocatable :: W

      real(kind=8), dimension(2,2) :: KMAT, EMAT

      integer, dimension(:), allocatable :: ICELNOD
      integer, dimension(:), allocatable :: INDEX, ITEM
      integer(kind=4) :: NEIBPETOT, BUFlength, PETOT
      integer(kind=4), dimension(2) :: NEIBPE

      integer(kind=4), dimension(0:2) :: import_index, export_index
      integer(kind=4), dimension(  2) :: import_item , export_item

      real(kind=8), dimension(2) :: SENDbuf, RECVbuf

      integer(kind=4), dimension(:,:), allocatable :: stat_send
      integer(kind=4), dimension(:,:), allocatable :: stat_recv
      integer(kind=4), dimension(:  ), allocatable :: request_send
      integer(kind=4), dimension(:  ), allocatable :: request_recv

!C
!C +-------+
!C | INIT. |
!C +-------+
!C===
!C
!C-- MPI init.
      call MPI_Init      (ierr)
      call MPI_Comm_size (MPI_COMM_WORLD, PETOT, ierr )
      call MPI_Comm_rank (MPI_COMM_WORLD, my_rank, ierr )

!C
!C-- CTRL data 
      if (my_rank.eq.0) then
        open  (11, file='input.dat', status='unknown')
         read (11,*) NEg
         read (11,*) dX, QV, AREA, COND
         read (11,*) ITERmax
         read (11,*) EPS
         write (*,'(i10)') NEg
        close (11)
      endif

      call MPI_Bcast (NEg    , 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierr)
      call MPI_Bcast (ITERmax, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierr)
      call MPI_Bcast (dX     , 1, MPI_DOUBLE_PRECISION, 0,              &
     &                                            MPI_COMM_WORLD, ierr)
      call MPI_Bcast (QV     , 1, MPI_DOUBLE_PRECISION, 0,              &
     &                                            MPI_COMM_WORLD, ierr)
      call MPI_Bcast (AREA   , 1, MPI_DOUBLE_PRECISION, 0,              &
     &                                            MPI_COMM_WORLD, ierr)
      call MPI_Bcast (COND   , 1, MPI_DOUBLE_PRECISION, 0,              &
     &                                            MPI_COMM_WORLD, ierr)
      call MPI_Bcast (EPS,   1, MPI_DOUBLE_PRECISION, 0,                &
     &                                            MPI_COMM_WORLD, ierr)

!C
!C-- Local Mesh Size
      Ng= NEg + 1
      N = Ng / PETOT

      nr = Ng - N*PETOT
      if (my_rank.lt.nr) N= N+1
  
      NE= N - 1 + 2
      NP= N + 2

      if (my_rank.eq.0) NE= N - 1 + 1
      if (my_rank.eq.0) NP= N + 1
      if (my_rank.eq.PETOT-1) NE= N - 1 + 1
      if (my_rank.eq.PETOT-1) NP= N + 1

      if (PETOT.eq.1) NE= N-1
      if (PETOT.eq.1) NP= N  

!C
!C-- ARRAYs
      allocate (PHI(NP), DIAG(NP), AMAT(2*NP-2), RHS(NP))
      allocate (ICELNOD(2*NE))
      allocate (INDEX(0:NP), ITEM(2*NP-2), W(NP,4))
       PHI= 0.d0
      AMAT= 0.d0
      DIAG= 0.d0
       RHS= 0.d0

      do icel= 1, NE
        ICELNOD(2*icel-1)= icel
        ICELNOD(2*icel  )= icel + 1
      enddo

      if (PETOT.gt.1) then
      if (my_rank.eq.0) then
	icel= NE
	ICELNOD(2*icel-1)= N
	ICELNOD(2*icel  )= N + 1
       else if (my_rank.eq.PETOT-1) then
	icel= NE
	ICELNOD(2*icel-1)= N + 1
	ICELNOD(2*icel  )= 1
       else
	icel= NE - 1
	ICELNOD(2*icel-1)= N + 1
	ICELNOD(2*icel  )= 1
	icel= NE
	ICELNOD(2*icel-1)= N
	ICELNOD(2*icel  )= N + 2
      endif
      endif

      KMAT(1,1)= +1.d0
      KMAT(1,2)= -1.d0
      KMAT(2,1)= -1.d0
      KMAT(2,2)= +1.d0
!C===

!C
!C +--------------+
!C | CONNECTIVITY |
!C +--------------+
!C===
      INDEX   = 2

      INDEX(0)= 0

      INDEX(N+1)= 1
      INDEX(NP )= 1

      if (my_rank.eq.0)       INDEX(1)= 1 
      if (my_rank.eq.PETOT-1) INDEX(N)= 1

      do i= 1, NP
        INDEX(i)= INDEX(i) + INDEX(i-1)      
      enddo

      NPLU= INDEX(NP)
      ITEM= 0

      do i= 1, N
        jS= INDEX(i-1)
        if (my_rank.eq.0.and.i.eq.1) then
          ITEM(jS+1)= i+1
         else if (my_rank.eq.PETOT-1.and.i.eq.N) then
          ITEM(jS+1)= i-1
         else
          ITEM(jS+1)= i-1            
          ITEM(jS+2)= i+1     
          if (i.eq.1) ITEM(jS+1)= N + 1
          if (i.eq.N) ITEM(jS+2)= N + 2
          if (my_rank.eq.0.and.i.eq.N) ITEM(jS+2)= N + 1
        endif
      enddo



      i = N + 1
      jS= INDEX(i-1)
      if (my_rank.eq.0) then
	ITEM(jS+1)= N
       else
	ITEM(jS+1)= 1
      endif

      i = N + 2
      if (my_rank.ne.0.and.my_rank.ne.PETOT-1) then
        jS= INDEX(i-1)
	ITEM(jS+1)= N
      endif


!C
!C-- COMMUNICATION
      NEIBPETOT= 2
      if (my_rank.eq.0      ) NEIBPETOT= 1
      if (my_rank.eq.PETOT-1) NEIBPETOT= 1
      if (PETOT.eq.1)         NEIBPETOT= 0

      NEIBPE(1)= my_rank - 1
      NEIBPE(2)= my_rank + 1

      if (my_rank.eq.0      ) NEIBPE(1)= my_rank + 1
      if (my_rank.eq.PETOT-1) NEIBPE(1)= my_rank - 1

      BUFlength= 1

      import_index= 0
      export_index= 0
      import_item = 0
      export_item = 0

      import_index(1)= 1
      import_index(2)= 2
      import_item (1)= N+1
      import_item (2)= N+2

      export_index(1)= 1
      export_index(2)= 2
      export_item (1)= 1
      export_item (2)= N

      if (my_rank.eq.0) then
        import_item (1)= N+1
        export_item (1)= N
      endif
!C
!C-- INIT. arrays for MPI_Waitall

      allocate (stat_send(MPI_STATUS_SIZE,2*NEIBPETOT))
      allocate (request_send(2*NEIBPETOT))
!C===
      call MPI_Barrier (MPI_COMM_WORLD, ierr)
      S1Time = MPI_Wtime()
!C
!C +-----------------+
!C | MATRIX ASSEMBLE |
!C +-----------------+
!C===
      do icel= 1, NE
        in1= ICELNOD(2*icel-1)
        in2= ICELNOD(2*icel  )
        DL = dX

        cK= AREA*COND/DL
        EMAT(1,1)= Ck*KMAT(1,1)
        EMAT(1,2)= Ck*KMAT(1,2)
        EMAT(2,1)= Ck*KMAT(2,1)
        EMAT(2,2)= Ck*KMAT(2,2)

        DIAG(in1)= DIAG(in1) + EMAT(1,1)
        DIAG(in2)= DIAG(in2) + EMAT(2,2)


        if (my_rank.eq.0.and.icel.eq.1) then
          k1= INDEX(in1-1) + 1  
         else
          k1= INDEX(in1-1) + 2  
        endif

        k2= INDEX(in2-1) + 1  

        AMAT(k1)= AMAT(k1) + EMAT(1,2)
        AMAT(k2)= AMAT(k2) + EMAT(2,1)

        QN= 0.50d0*QV*AREA*DL
        RHS(in1)= RHS(in1) + QN
        RHS(in2)= RHS(in2) + QN
      enddo
!C===

!C
!C +---------------------+
!C | BOUNDARY CONDITIONS |
!C +---------------------+
!C===

!C
!C-- X=Xmin
      if (my_rank.eq.0) then
        i = 1
        jS= INDEX(i-1)

        AMAT(jS+1)= 0.d0
        DIAG(i)= 1.d0
        RHS (i)= 0.d0
        do k= 1, NPLU
          if (ITEM(k).eq.1) AMAT(k)= 0.d0
        enddo
      endif
!C===
      E1Time = MPI_Wtime()

!C
!C +---------------+
!C | CG iterations |
!C +---------------+
!C===
      R = 1
      Z = 2
      Q = 2
      P = 3
      DD= 4

      do i= 1, N
        W(i,DD)= 1.0D0 / DIAG(i)
      enddo

!C
!C-- {r0}= {b} - [A]{xini} |

!C-    init
        do neib= 1, NEIBPETOT
          do k= export_index(neib-1)+1, export_index(neib)
            kk= export_item(k)
            SENDbuf(k)= PHI(kk)
          enddo  
        enddo  

!C
!C-- SEND & RECV.
      do neib= 1, NEIBPETOT
        is   = export_index(neib-1) + 1
        len_s= export_index(neib) - export_index(neib-1) 
        call MPI_Isend (SENDbuf(is), len_s, MPI_DOUBLE_PRECISION,       &
     &                  NEIBPE(neib), 0, MPI_COMM_WORLD,                &
     &                  request_send(neib), ierr)
      enddo

      do neib= 1, NEIBPETOT
        ir   = import_index(neib-1) + 1
        len_r= import_index(neib) - import_index(neib-1) 
        call MPI_Irecv (PHI(ir+N), len_r, MPI_DOUBLE_PRECISION,         &
     &                  NEIBPE(neib), 0, MPI_COMM_WORLD,                &
     &                  request_send(neib+NEIBPETOT), ierr)
      enddo
      call MPI_Waitall (2*NEIBPETOT, request_send, stat_send, ierr)

      do i= 1, N
        W(i,R) = DIAG(i)*PHI(i)
        do j= INDEX(i-1)+1, INDEX(i)
          W(i,R) = W(i,R) + AMAT(j)*PHI(ITEM(j))
        enddo
      enddo

      BNRM20= 0.0D0
      do i= 1, N
        BNRM20 = BNRM20  + RHS(i)  **2
        W(i,R) = RHS(i) - W(i,R)
      enddo
      call MPI_Allreduce (BNRM20, BNRM2, 1, MPI_DOUBLE_PRECISION,       &
     &                    MPI_SUM, MPI_COMM_WORLD, ierr)

!C********************************************************************
      do iter= 1, ITERmax
!C
!C-- {z}= [Minv]{r}

      do i= 1, N
        W(i,Z)= W(i,DD) * W(i,R)
      enddo

!C
!C-- RHO= {r}{z}

      RHO0= 0.d0
      do i= 1, N
        RHO0= RHO0 + W(i,R)*W(i,Z)   
      enddo     
      call MPI_Allreduce (RHO0, RHO, 1, MPI_DOUBLE_PRECISION,           &
     &                    MPI_SUM, MPI_COMM_WORLD, ierr)

!C
!C-- {p} = {z} if      ITER=1  
!C   BETA= RHO / RHO1  otherwise 

      if ( iter.eq.1 ) then
        do i= 1, N
          W(i,P)= W(i,Z)
        enddo
       else
         BETA= RHO / RHO1
         do i= 1, N
           W(i,P)= W(i,Z) + BETA*W(i,P)
         enddo
      endif

!C
!C-- {q}= [A]{p}

!C-    init
        do neib= 1, NEIBPETOT
          do k= export_index(neib-1)+1, export_index(neib)
            kk= export_item(k)
            SENDbuf(k)= W(kk,P)
          enddo  
        enddo  

!C
!C-- SEND & RECV.
      do neib= 1, NEIBPETOT
        is   = export_index(neib-1) + 1
        len_s= export_index(neib) - export_index(neib-1) 
        call MPI_Isend (SENDbuf(is), len_s, MPI_DOUBLE_PRECISION,       &
     &                  NEIBPE(neib), 0, MPI_COMM_WORLD,                &
     &                  request_send(neib), ierr)
      enddo

      do neib= 1, NEIBPETOT
        ir   = import_index(neib-1) + 1
        len_r= import_index(neib) - import_index(neib-1) 
        call MPI_Irecv (W(ir+N,P), len_r, MPI_DOUBLE_PRECISION,         &
     &                  NEIBPE(neib), 0, MPI_COMM_WORLD,                &
     &                  request_send(neib+NEIBPETOT), ierr)
      enddo
      call MPI_Waitall (2*NEIBPETOT, request_send, stat_send, ierr)

      do i= 1, N
        W(i,Q) = DIAG(i)*W(i,P)
        do j= INDEX(i-1)+1, INDEX(i)
          W(i,Q) = W(i,Q) + AMAT(j)*W(ITEM(j),P)
        enddo
      enddo

!C
!C-- ALPHA= RHO / {p}{q}

      C10= 0.d0
      do i= 1, N
        C10= C10 + W(i,P)*W(i,Q)
      enddo

      call MPI_Allreduce (C10, C1, 1, MPI_DOUBLE_PRECISION,             &
     &                    MPI_SUM, MPI_COMM_WORLD, ierr)

      ALPHA= RHO / C1

!C
!C-- {x}= {x} + ALPHA*{p}
!C   {r}= {r} - ALPHA*{q}

      do i= 1, N
        PHI(i)= PHI(i) + ALPHA * W(i,P)
        W(i,R)= W(i,R) - ALPHA * W(i,Q)
      enddo

      DNRM20 = 0.0
      do i= 1, N
        DNRM20= DNRM20 + W(i,R)**2
      enddo

      call MPI_Allreduce (DNRM20, DNRM2, 1, MPI_DOUBLE_PRECISION,       &
     &                    MPI_SUM, MPI_COMM_WORLD, ierr)

        RESID= dsqrt(DNRM2/BNRM2)

        if (my_rank.eq.0.and.mod(iter,1000).eq.0) then
          write (*, '(i8,1pe16.6)') iter, RESID
        endif

        if ( RESID.le.EPS) goto 900
        RHO1 = RHO

      enddo
!C********************************************************************

      IER = 1

  900 continue

      E2Time = MPI_Wtime()

        if (my_rank.eq.0) then
          write (*, '(i8,1pe16.6)') iter, RESID
        endif
!C===

!C
!C-- OUTPUT
        if (my_rank.eq.0) then
          write (*,'(2(1pe16.6),a)') E1Time-S1Time, E2Time-E1Time, 
     &                               'sec.'
        endif


      if (my_rank.eq.PETOT-1) then
        write (*,'(/a)') '### TEMPERATURE'
        write (*,'(2i8, 1pe27.20 //)') my_rank, N, PHI(N)
      endif

      call MPI_FINALIZE (ierr)
      end program heat1Dp
